### PR TITLE
Feature/countable stream iterator

### PR DESCRIPTION
**Pull request by basz**
Just to get some eyes. Would something like this be able to get merged? I realise this might be a BC on for anyone using the PdoStreamIterator manually.

Problem: Unable to do `iterator_count($stream)` inside a loop as that will move the pointer (in unexpected ways). This prevents me from creating a replay script that properly gives feedback of the events to be processed when events are added to the stream after the loop has started.

```
foreach ($stream as $messages) {
    $count = iterator_count($stream); // will move pointer to end of stream.
}
```

So this PR implements a simple countable on the PdoStreamIterator. For this to work a countStatement must be passed from the *EventStore::load method (where the PdoStreamIterator is created)

### CODE REVIEW

File: src/PdoStreamIterator.php [hunk 1]
```diff
@@ -238,9 +261,17 @@ private function buildStatement(int $fromNumber): PDOStatement
...
+        $this->selectStatement->bindValue(':fromNumber', $fromNumber, PDO::PARAM_INT);
+        $this->selectStatement->bindValue(':limit', $limit, PDO::PARAM_INT);
+
+        return $this->selectStatement;
+    }
+
+    private function buildCountStatement(int $fromNumber): PDOStatement
+    {
+        $this->countStatement->bindValue(':fromNumber', $fromNumber, PDO::PARAM_INT);
+        $this->countStatement->bindValue(':limit', PHP_INT_MAX, PDO::PARAM_INT);
```
[the remark this answers is not in the document]
Comment (basz) [reply src/PdoStreamIterator.php]: agreed.

File: src/PdoStreamIterator.php [hunk 2]
```diff
@@ -76,14 +82,16 @@
...
+        PDOStatement $queryStatement,
         MessageFactory $messageFactory,
         int $batchSize,
         int $fromNumber,
         ?int $count,
         bool $forward
     ) {
-        $this->statement = $statement;
+        $this->selectStatement = $selectStatement;
+        $this->queryStatement = $queryStatement;
```
[the remark this answers is not in the document]
Comment (basz) [reply src/PdoStreamIterator.php]: just a starting point, if you are willing to consider this for acceptance I will make it look nice and perfect.

File: src/PdoStreamIterator.php [hunk 3]
```diff
@@ -76,14 +82,16 @@
     private $forward;
 
     public function __construct(
-        PDOStatement $statement,
+        PDOStatement $selectStatement,
+        PDOStatement $queryStatement,
```
[the remark this answers is not in the document]
Comment (basz) [reply src/PdoStreamIterator.php]: Regarding the loop. If you look closely I only update the Progress every 5 seconds and that would be the place where I would do a count on the iterator. That means that it is still possible to have 101% displayed or "1001 of 1000" Perhaps I should change it that when there are only a few seconds left to go I to that more often... But with this PR I *could* do that...

[the remark this answers is not in the document]
Comment (basz) [reply src/PdoStreamIterator.php]: here the exact utilisation of countable https://gist.github.com/basz/36fcf387f5b7cb76dc82a997eb3bddca#file-gistfile1-txt-L181

[the remark this answers is not in the document]
Comment (prolic) [reply src/PdoStreamIterator.php]: As far as I am concerned, showing 101% is not a real problem. It simply means there was a little more work to do then initially expected.

As for this PR: We only promise to return an instance of Iterator. Therefore the PdoStreamIterator is only an implementation detail. You can't expect the same behavior for every event store implementation, but if you know you only connect to PDO, I think it's a valuable addition. If you can provide some docs and how to use it and list some limitations (as discussed: dependend on PDO only, could show 101%, etc), i'll merge it.

[the remark this answers is not in the document]
Comment (basz) [reply src/PdoStreamIterator.php]: agree on 101%

We *could* promise to return a `(Prooph)StreamIterator` that implements Iterator and Countable. No BC.

Should go into `prooph/event-store` then the `EmptyIterator` should be moved there then too.

What do you think?

**Comment by prolic** [comment]
Sure, if you want to do this work no problem

**Comment by prolic** [comment]
@basz any status update?

**Comment by basz** [comment]
I’ve been a bit distracted

I’ve implemented against the PR on event-store (see inmemorystore) and got a shitload of assertCount failures that I need to investigate. It *might* be that assertCount uses count_iterator which I now know messes with the pointer. Maybe your tests weren’t correct to begin with but the metadata selectors are a bit hard to follow... :)

So not sure - but not dead. Maybe in de weekend.

**Comment by basz** [comment]
@prolic I think I got the implemented complete.

some event-store concerns;

1. I'm not sure I understand why we would need to return an Empty(Stream)Iterator instance and not simply use the PdoStreamIterator/InMemoryStreamIterator as is? An advantage would be that a user could keep the (empty) StreamIterator around, wait a while rewind it and try again...
To me it seems [this](https://github.com/prooph/pdo-event-store/pull/172/commits/6bf2ed069cc809dded7d14bac9060aa3e62ef073) duplicates [this](https://github.com/prooph/pdo-event-store/pull/172/commits/a5062a3f090fedd01dd2341b1f6891d1f855192e). The only reason I can think of is that users might check for `instance EmptyIterator` (most likely in their tests).

1.1. Couldn't we return `new ArrayStreamIterator([])` as an alternative?

2. I found an a query to detect the existence of the stream. It  *only* [executes](https://github.com/prooph/pdo-event-store/blob/master/src/PostgresEventStore.php#L237-L243) this on `PostgresEventStore::load`. Is this forgotten logic which can be removed?

**Comment by prolic** [comment]
The empty iterator is just an implementation detail. It can be anything else as well. I have reservations about ArrayIterator, because it's not immutable and you can append to the stream, which is not what we want.

load-method needs to check for stream existence, because otherwise you'll have SQL errors during querying, which can cause problems during longer transaction (they are rolled back then),

**Comment by basz** [comment]
> load-method needs to check for stream existence, because otherwise you'll have SQL errors during querying, which can cause problems during longer transaction (they are rolled back then),

sure, but (sorry) why not for loadReverse too and for all vendors then? Also there is an hasStream method which does the same thing (queries a row for real_stream_name)?

(but out of scope of this PR i guess, please go ahead and review)

**Comment by prolic** [comment]
Really? Damn it, then there is probably another bug hidden noone found so
far. I'll have to check this, thanks for pointing me to this.

I'll try to review your changes asap.

Sascha-Oliver Prolic


Am So., 23. Sep. 2018 um 17:08 Uhr schrieb Bas Kamer <
notifications@github.com>:

**Comment by prolic** [comment]
@basz I thought maybe we can get https://github.com/prooph/pdo-event-store/pull/175 done first.
